benchmarks/jdk-streams/src/main/scala/org/renaissance/jdk/streams/Scrabble.scala

```diff
@@ -1,59 +1,66 @@
 @Summary("Solves the Scrabble puzzle using JDK Streams.")
 @Licenses(Array(License.GPL2))
 @Repetitions(50)
-class Scrabble extends RenaissanceBenchmark {
-
-  class ScrabbleResult(actualResult: List[Entry[Integer, List[String]]], expected: Seq[String])
-    extends BenchmarkResult {
-
-    override def validate(): Unit = {
-      val actualWords = JavaScrabble.prepareForValidation(actualResult)
-      ValidationException.throwIfNotEqual(expected.size, actualWords.size, "best words count")
-
-      for ((expected, actual) <- expected zip JavaConverters.asScalaBuffer(actualWords)) {
-        ValidationException.throwIfNotEqual(expected, actual, "best words")
-      }
-    }
-  }
+@Parameter(name = "input_path", defaultValue = "/shakespeare.txt")
+@Parameter(
+  name = "expected_result",
+  defaultValue = "120--QUICKLY,118--ZEPHYRS,114--QUALIFY-QUICKEN-QUICKER"
+)
+@Configuration(
+  name = "test",
+  settings = Array(
+    "input_path = /shakespeare-truncated.txt",
+    "expected_result = 120--QUICKLY,114--QUICKEN-QUICKER,108--BLAZING-PRIZING"
+  )
+)
+@Configuration(name = "jmh")
+final class Scrabble extends Benchmark {
 
   // TODO: Consolidate benchmark parameters across the suite.
   //  See: https://github.com/renaissance-benchmarks/renaissance/issues/27
 
-  var shakespearePath = "/shakespeare.txt"
+  private var inputPathParam: String = _
 
-  var scrabblePath = "/scrabble.txt"
+  private var expectedResultParam: Seq[String] = _
 
-  var scrabble: JavaScrabble = null
+  private val scrabblePath = "/scrabble.txt"
 
-  val expectedResultFull = Seq("120--QUICKLY", "118--ZEPHYRS", "114--QUALIFY-QUICKEN-QUICKER")
-  val expectedResultTest = Seq("120--QUICKLY", "114--QUICKEN-QUICKER", "108--BLAZING-PRIZING")
+  private var scrabble: JavaScrabble = _
 
-  override def setUpBeforeAll(c: Config): Unit = {
-    if (c.functionalTest) {
-      shakespearePath = "/shakespeare-truncated.txt"
-    }
-    scrabble = new JavaScrabble(shakespearePath, scrabblePath)
+  override def setUpBeforeAll(c: BenchmarkContext): Unit = {
+    inputPathParam = c.stringParameter("input_path")
+    expectedResultParam = c.stringParameter("expected_result").split(",").map(_.trim).toSeq
+    scrabble = new JavaScrabble(inputPathParam, scrabblePath)
   }
 
-  override def runIteration(c: Config): BenchmarkResult = {
+  override def runIteration(c: BenchmarkContext): BenchmarkResult = {
     val result = scrabble.run()
-    return new ScrabbleResult(
-      result,
-      if (c.functionalTest) expectedResultTest else expectedResultFull
-    )
+
+    () => {
+      val actualWords = JavaScrabble.prepareForValidation(result)
+      BenchmarkResult.assertEquals(
+        expectedResultParam.size,
+        actualWords.size,
+        "best words count"
+      )
+
+      for ((expected, actual) <- expectedResultParam zip JavaConverters.asScalaBuffer(
+             actualWords
+           )) {
+        BenchmarkResult.assertEquals(expected, actual, "best words")
+      }
+    }
```

Comment:
I think we should enforce the use of `return` in Scala sources. As I was switching between Java and Scala sources, my first impression was *we should not do validation inside `runIteration`*. But the missing `return` *probably* means that we return anonymous implementation of `BenchmarkResult` and the code inside `{ ... }` would be executed later. Wasn't a separate class actually more readable?

Just my two (despairing) cents.

Reply:
I don't know about `return` enforcement. Leaving it out seems to be an accepted practice and I was annoyed by the various yellow markers in IDEA. If a method has uncomplicated control flow (like this one), I did not find it too confusing or annoying once I got used to Scala. However, if a method has an early return somewhere, I would probably prefer to see explicit returns on all paths.

And yes, this returns an anonymous implementation of `BenchmarkResult`. It is a bit more verbose, because this particular Scala version does not support using a lambda implementation of a SAM class (Single Anonymous Method).

As for separate class, I was removing simple classes that only serve to delay the validation computation and can be easily put into a lambda.

Reply:
I think that the missing `return` is a bad (despite being accepted) practice but okay :-). We are still several characters away from PERL syntax so let's keep it like this.